### PR TITLE
Add clients and client view

### PR DESCRIPTION
**Pull request by helfi92**

### CODE REVIEW

**Comment by helfi92** [comment]
### Clients

screen shot 2018-05-02 at 4 27 28 pm

---

### Edit/View Client

screen shot 2018-05-02 at 4 28 01 pm

---

### Create Client

screen shot 2018-05-02 at 4 31 34 pm

File: src/views/Clients/ViewClients/index.jsx [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+                />
+              </FormGroup>
+              <ClientsTable
+                onPageChange={this.handlePageChange}
+                clientsConnection={clients}
+                sortByLastUsed={this.state.sortByLastUsed}
+              />
+            </Fragment>
+          )}
+          <Button
```
Comment (eliperelman) [inline src/views/Clients/ViewClients/index.jsx]: Speed dial?

File: src/views/Clients/index.jsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+const ViewClients = loadable(() =>
+  import(/* webpackChunkName: 'Clients.ViewClients' */ './ViewClients')
+);
+const ViewClient = loadable(() =>
+  import(/* webpackChunkName: 'Clients.ViewClient' */ './ViewClient')
+);
+
+@hot(module)
+export default class Client extends Component {
```
Comment (eliperelman) [inline src/views/Clients/index.jsx]: Class name should be "Clients".

File: src/views/Clients/index.jsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    const {
+      match: { path },
+      ...props
+    } = this.props;
+
+    return (
+      <Switch>
+        <RouteWithProps
+          path={`${path}/create`}
+          newClient
```
Comment (eliperelman) [inline src/views/Clients/index.jsx]: Rename to `isNew` or `isNewClient`.

File: src/views/Clients/ViewClients/index.jsx [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    return (
+      <Dashboard
+        title="Clients"
+        search={
+          <Search
+            disabled={loading}
+            value={clientSearch}
+            onChange={this.handleClientSearchChange}
+            onSubmit={this.handleClientSearchSubmit}
+            placeholder="Client beginning with"
```
Comment (eliperelman) [inline src/views/Clients/ViewClients/index.jsx]: Let's rename this placeholder so it's visible when the search bar is shrunk.

File: src/components/ClientsTable/index.jsx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            sortByLastUsed,
+            clientsConnection
+          )}
+          pageSize={VIEW_CLIENTS_PAGE_SIZE}
+          columnsSize={1}
+          onPageChange={onPageChange}
+          renderRow={({ node: client }) => (
+            <TableRow key={client.clientId}>
+              <TableCell>{client.clientId}</TableCell>
+              <TableCell>
```
[the remark this answers is not in the document]
Comment (helfi92) [reply src/components/ClientsTable/index.jsx]: That's a good point. I don't know why I felt the need to use the pencil icon. I will revise.

File: src/components/DateChooser/index.jsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import MuiPickersUtilsProvider from 'material-ui-pickers/utils/MuiPickersUtilsProvider';
+import ChevronRightIcon from 'mdi-react/ChevronRightIcon';
+import CalendarIcon from 'mdi-react/CalendarIcon';
+import ChevronLeftIcon from 'mdi-react/ChevronLeftIcon';
+import { date } from '../../utils/prop-types';
+
+/**
+ * Display a date picker modal to select a date.
+ */
+export default class DateChooser extends Component {
```
[the remark this answers is not in the document]
Comment (helfi92) [reply src/components/DateChooser/index.jsx]: I like that.

File: src/utils/splitLines.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { uniq, pipe, map, filter, split } from 'ramda';
+
+/**
+ * Splits a string into an array of strings using newlines as a separator.
+ * Returns an array with unique values.
+ */
+export default pipe(
+  split(/[\r\n]+/),
+  map(line => line.trim()),
```
[the remark this answers is not in the document]
Comment (helfi92) [reply src/utils/splitLines.js]: 😍

File: src/components/ClientForm/index.jsx [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+import ContentSaveIcon from 'mdi-react/ContentSaveIcon';
+import LinkIcon from 'mdi-react/LinkIcon';
+import LockResetIcon from 'mdi-react/LockResetIcon';
+import SpeedDial from '../SpeedDial';
+import DateChooser from '../DateChooser';
+import { client } from '../../utils/prop-types';
+// import splitLines from '../../utils/splitLines';
+
+@withStyles(theme => ({
+  saveButton: {
```
[the remark this answers is not in the document]
Comment (helfi92) [reply src/components/ClientForm/index.jsx]: The speed dial already has these styles. This style is being applied to a button fab.

[the remark this answers is not in the document]
Comment (eliperelman) [reply src/components/ClientForm/index.jsx]: Right, why not just use the speed dial?

File: src/components/ClientForm/index.jsx [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+                    </List>
+                  }
+                />
+              </ListItem>
+            </Fragment>
+          ) : null}
+        </List>
+        {newClient ? (
+          <Tooltip title="Save">
+            <Button
```
[the remark this answers is not in the document]
Comment (eliperelman) [reply src/components/ClientForm/index.jsx]: If the speed dial can act just like a fab, then at least with the speed dial we get the fab styles and toolip functionality.

[the remark this answers is not in the document]
Comment (helfi92) [reply src/components/ClientForm/index.jsx]: Are you suggesting that we change the behavior of the speed dial component to act like a fab button when there is only 1 action?

[the remark this answers is not in the document]
Comment (eliperelman) [reply src/components/ClientForm/index.jsx]: Maybe, does it not work that way right now?

**Comment by helfi92** [comment]
Added a new mixin for fab positioning. Let me know if it's okay now.

**Comment by eliperelman** [comment]
@helfi92 can you do this from `overrides` instead of mixins?

https://material-ui-next.com/customization/themes/#customizing-all-instances-of-a-component-type

Basically, override the styles of `MuiButton`'s `fab` styles.

**Comment by helfi92** [comment]
@eliperelman Regarding `overrides`, we would have to do

```
overrides: {
  MuiButton: {
    fab: {
      position: 'fixed',
      bottom: 16,
      right: 24, 
      
      borderRadius: '50%',
      padding: 0,
      minWidth: 0,
      width: 56,
      fontSize: 24,
      height: 56,
      boxShadow: theme.shadows[6],
      '&:active': {
        boxShadow: theme.shadows[12],
      },
    }
  }  
}
```
https://github.com/mui-org/material-ui/blob/v1-beta/packages/material-ui/src/Button/Button.js#L118-L128

If Material-UI decides to alter these values in the future, we might need to change the values on our end as well. That being said, it might be safer to stay with `mixins`. Thoughts?

**Comment by eliperelman** [comment]
Isn't the override a merge operation?

**Comment by helfi92** [comment]
@eliperelman  Hm, maybe it is..but I realized that setting `position: fixed` to all fab buttons defeats the whole purpose of the SpeedDial component.  This will collapse all of the speed dial actions and have them sit on top of each other.

screen shot 2018-05-07 at 3 38 40 pm

**Comment by eliperelman** [comment]
lol good point. We can either do the mixin or have a fab component.

**Comment by helfi92** [comment]
Ok. I think using mixins should be okay. We can always change later.

**Pull request merged**